hare/messagevalidation.go

```diff
@@ -103,30 +103,48 @@ func (validator *syntaxContextValidator) ContextuallyValidateMessage(m *Msg, cur
 		return false, errors.New("nil inner message")
 	}
 
-	// PreRound & Notify are always contextually valid
+	// first validate pre-round and notify
 	switch m.InnerMsg.Type {
-	case PreRound:
+	case Pre:
 		return true, nil
-	case Proposal:
-		if currentK == Round1 {
-			return false, errEarlyMsg
-		}
-		if currentK == Round2 || currentK == Round3 {
-			return true, nil // process proposals for
-		}
-		return false, nil
 	case Notify:
 		return true, nil
 	}
 
-	// Status & Commit Messages should match the expected K
-	if currentK == m.InnerMsg.K { // arrived on time
-		return true, nil
+	// the message must match the current iteration unless it is a notify message
+	currentIteration := currentK / 4
+	msgIteration := m.InnerMsg.K / 4
+	if currentIteration != msgIteration {
+		return false, nil
```

Comment:
you must return an error (at least for the sake of printing the reason why this validation failed)

Reply:
Fixed